wdl-analysis/src/types/v1.rs

```diff
@@ -1004,6 +1046,472 @@ where
         }
     }
 
+    /// Evaluates a `runtime` section item.
+    pub(crate) fn evaluate_runtime_item(
+        &mut self,
+        scope: &ScopeRef<'_>,
+        name: &Ident,
+        expr: &Expr,
+    ) {
+        let expr_ty = self.evaluate_expr(scope, expr).unwrap_or(Type::Union);
+        if !self.evaluate_requirement(name, expr, expr_ty) {
+            // Always use object types for `runtime` section `inputs` and `outputs` keys as
+            // only `hints` sections can use input/output hidden types
+            self.evaluate_hint(name, expr, expr_ty, true);
+        }
+    }
+
+    /// Evaluates a `requirements` section item.
+    pub(crate) fn evaluate_requirements_item(
+        &mut self,
+        scope: &ScopeRef<'_>,
+        name: &Ident,
+        expr: &Expr,
+    ) {
+        let expr_ty = self.evaluate_expr(scope, expr).unwrap_or(Type::Union);
+        self.evaluate_requirement(name, expr, expr_ty);
+    }
+
+    /// Evaluates a requirement in either a `requirements` section or a legacy
+    /// `runtime` section.
+    ///
+    /// Returns `true` if the name matched a requirement or `false` if it did
+    /// not.
+    fn evaluate_requirement(&mut self, name: &Ident, expr: &Expr, expr_ty: Type) -> bool {
+        match name.as_str() {
+            "container" | "docker" => {
+                if !expr_ty.is_coercible_to(self.types, &PrimitiveTypeKind::String.into())
+                    && !expr_ty.is_coercible_to(self.types, &STDLIB.array_string)
+                {
+                    self.diagnostics.push(type_mismatch_custom(
+                        self.types,
+                        "type `String` or type `Array[String]`",
+                        name.span(),
+                        expr_ty,
+                        expr.span(),
+                    ));
+                }
+
+                true
+            }
+            "cpu" => {
+                if !expr_ty.is_coercible_to(self.types, &PrimitiveTypeKind::Float.into()) {
+                    self.diagnostics.push(type_mismatch_custom(
+                        self.types,
+                        "type `Int` or type `Float`",
+                        name.span(),
+                        expr_ty,
+                        expr.span(),
+                    ));
+                }
+
+                true
+            }
+            "memory" => {
+                if !expr_ty.is_coercible_to(self.types, &PrimitiveTypeKind::Integer.into())
+                    && !expr_ty.is_coercible_to(self.types, &PrimitiveTypeKind::String.into())
+                {
+                    self.diagnostics.push(type_mismatch_custom(
+                        self.types,
+                        "type `Int` or type `String`",
```

Comment:
At what point will `String` entries be checked for conforming to the expected format? e.g. they use the proper units etc.

Reply:
That's a good question; while some of these values could be statically validated, such as having a proper suffix, some of the values will need to be validated by the executor.

I think we could probably do a follow-up commit to add suffix validation for static analysis, at least.

Reply:
Oh as these are expressions and may depend on user inputs, there's no way to statically validate these other than when they are specified as string literals; validation of these values should probably be delayed until evaluation.